‘const llvm::AttributeList’ passed to llvm::AttributeList::addAttributes where ‘const llvm::AttrBuilder&’ expected in codegen.cpp

Build failed with the following error when Julia was at 1eef0279982c3b8b923ff9cd78bea6cfc07fb22c and LLVM was at https://github.com/llvm-mirror/llvm/commit/74fe5ca7247ea8b37fd61ef79faaec0bfac1a374
```
In file included from /home/sanjay/Software/julia_vanilla/julia_src/src/intrinsics.cpp:7:0,
                 from /home/sanjay/Software/julia_vanilla/julia_src/src/codegen.cpp:2209:
/home/sanjay/Software/julia_vanilla/julia_src/src/ccall.cpp: In member function ‘std::string function_sig_t::generate_func_sig()’:
/home/sanjay/Software/julia_vanilla/julia_src/src/ccall.cpp:1408:76: error: no matching function for call to ‘llvm::AttributeList::addAttributes(llvm::LLVMContext&, size_t, const llvm::AttributeList&)’
             attributes = attributes.addAttributes(jl_LLVMContext, i + 1, as);
                                                                            ^
/home/sanjay/Software/julia_vanilla/julia_src/src/ccall.cpp:1408:76: note: candidate is:
In file included from /home/sanjay/Software/julia_vanilla/llvm-svn_src/include/llvm/IR/Argument.h:19:0,
                 from /home/sanjay/Software/julia_vanilla/llvm-svn_src/include/llvm/IR/Function.h:25,
                 from /home/sanjay/Software/julia_vanilla/llvm-svn_src/include/llvm/Analysis/TargetLibraryInfo.h:16,
                 from /home/sanjay/Software/julia_vanilla/julia_src/src/codegen.cpp:48:
/home/sanjay/Software/julia_vanilla/llvm-svn_src/include/llvm/IR/Attributes.h:365:17: note: llvm::AttributeList llvm::AttributeList::addAttributes(llvm::LLVMContext&, unsigned int, const llvm::AttrBuilder&) const
   AttributeList addAttributes(LLVMContext &C, unsigned Index,
                 ^
/home/sanjay/Software/julia_vanilla/llvm-svn_src/include/llvm/IR/Attributes.h:365:17: note:   no known conversion for argument 3 from ‘const llvm::AttributeList’ to ‘const llvm::AttrBuilder&’
make[1]: *** [codegen.dbg.obj] Error 1
make: *** [julia-src-debug] Error 2
```

--- a/src/ccall.cpp
+++ b/src/ccall.cpp
@@ -1275,7 +1275,7 @@ std::string generate_func_sig()
     assert(rt && !jl_is_abstract_ref_type(rt));
 
 #if JL_LLVM_VERSION >= 50000
-    std::vector<AttributeList> paramattrs;
+    std::vector<AttrBuilder> paramattrs;
 #else
     std::vector<AttributeSet> paramattrs;
 #endif
@@ -1302,7 +1302,7 @@ std::string generate_func_sig()
 #endif
             retattrs.addAttribute(Attribute::NoAlias);
 #if JL_LLVM_VERSION >= 50000
-            paramattrs.push_back(AttributeList::get(jl_LLVMContext, 1, retattrs));
+            paramattrs.push_back(std::move(retattrs));
 #else
             paramattrs.push_back(AttributeSet::get(jl_LLVMContext, 1, retattrs));
 #endif
@@ -1388,24 +1388,26 @@ std::string generate_func_sig()
         do { // for each arg for which this type applies, add the appropriate LLVM parameter attributes
             if (i < nargs) { // if vararg, the last declared arg type may not have a corresponding arg value
 #if JL_LLVM_VERSION >= 50000
-                AttributeList params = AttributeList::get(jl_LLVMContext, i + sret + 1, ab);
+                paramattrs.push_back(std::move(ab));
 #else
                 AttributeSet params = AttributeSet::get(jl_LLVMContext, i + sret + 1, ab);
-#endif
                 paramattrs.push_back(params);
+#endif
             }
             i++;
         } while (current_isVa && i < nargs); // if is this is the vararg, loop to the end
     }
 
     for (i = 0; i < nargs + sret; ++i) {
+        const auto &as = paramattrs.at(i);
 #if JL_LLVM_VERSION >= 50000
-        const AttributeList &as = paramattrs.at(i);
+        if (!as.hasAttributes())
+            continue;
 #else
-        const AttributeSet &as = paramattrs.at(i);
+        if (as.isEmpty())
+            continue;
 #endif
-        if (!as.isEmpty())
-            attributes = attributes.addAttributes(jl_LLVMContext, i + 1, as);
+        attributes = attributes.addAttributes(jl_LLVMContext, i + 1, as);
     }
     if (rt == jl_bottom_type) {
         attributes = attributes.addAttribute(jl_LLVMContext,

--- a/src/cgmemmgr.cpp
+++ b/src/cgmemmgr.cpp
@@ -732,8 +732,11 @@ public:
     }
     void registerEHFrames(uint8_t *Addr, uint64_t LoadAddr,
                           size_t Size) override;
+#if 0
+    // Disable for now since we are not actually using this.
     void deregisterEHFrames(uint8_t *Addr, uint64_t LoadAddr,
                             size_t Size) override;
+#endif
     uint8_t *allocateCodeSection(uintptr_t Size, unsigned Alignment,
                                  unsigned SectionID,
                                  StringRef SectionName) override;
@@ -860,12 +863,14 @@ void RTDyldMemoryManagerJL::registerEHFrames(uint8_t *Addr,
     }
 }
 
+#if 0
 void RTDyldMemoryManagerJL::deregisterEHFrames(uint8_t *Addr,
                                                uint64_t LoadAddr,
                                                size_t Size)
 {
     deregister_eh_frames((uint8_t*)LoadAddr, Size);
 }
+#endif
 
 }
 

